### DISCUSSION TITLE
Dithering, etc.

### DISCUSSION THREAD

**Discussion by DevEvV** [arkypita/LaserGRBL#1330] [Ideas]
I have some suggestions and questions.

1. Fake 1-bit dithering. What does it mean? I'm going to talk about light dots on black because I'm engraving on stone. We are dithering the image. And then darken the points depending on how dark these points were in the original image. This method of engraving gives a much higher quality image on stone and glass. See the example only at a scale of 1: 1, for this you may need to download the image and open it in Photoshop.

![3](https://user-images.githubusercontent.com/78816441/107413331-b2a23600-6b21-11eb-8a79-d4ae7d279157.jpg)
![1](https://user-images.githubusercontent.com/78816441/107414771-87b8e180-6b23-11eb-927e-da450701a03f.jpg)

2. Zoom 1: 1 is required when importing images in 1-bit mode! Without it, it is impossible to estimate how the points will be located on important areas of the image. This is very important for those who constantly engrave portraits.

3. The new UsbSerial2 is much faster than the previous version. But if I create a program in LaserGRBL and then run it in Universal Gcode Sender Classic, it runs twice as fast. In the UGS Platform, it is not as fast as in the Classic version. The X-axis on my machine moves at a speed of 1000mm / s, but during the execution of the program the speed is much slower. Changes in acceleration do not affect the engraving speed in any way. ESP32 controller. Can LaserGRBL get the same speed as UGS Classic?

4. In the Russian version of the program, the word "disconnect" is spelled incorrectly. It says "Отключитьсz", but it should be "Отключиться". 

5. How to make X336.99 S0 be written instead of G1 X336.99 Y24.646 S0, And G1 commands are written only when the Y coordinate changes? Is it possible?
![5](https://user-images.githubusercontent.com/78816441/107418354-999c8380-6b27-11eb-839e-9f4ae79890d7.jpg)

6. It would be cool if instead of inverting the image, we could choose whether we engrave with black dots on white or vice versa. And only the program would change, not the image. And the image window would have the original picture. And so if we engrave on a stone we do not see that we are engraving ..

Sorry for my English

**Comment by lorrainerose** [comment]
i updated and now my line to line is option is gone

**Reply by arkypita** [reply]
This is not because you did the update. It's because you changed the configuration by disabling PWM support.

Please, if you have any requests for help, open new discussions and do not post under other threads than talking about other things.

**Reply by lorrainerose** [reply]
thank you

**Comment by arkypita** [comment]
@DevEvV 

> Fake 1-bit dithering

Sounds interesting. If I understand correctly it is a mix between dithering and grayscale. You have to see how it behaves in the "real world" (different materials and different types of lasers).
You can use the new "bypass" feature to generate the image with photoshop and process it in LaserGRBL without further processing.

I'd like to see some comparison photos on the finished product.

> Zoom 1: 1 is required when importing images in 1-bit mode!

Yes, I know... vectorization would also be beneficial of a 1:1 preview... but there are problems doing that!

First of all the preview has to be fast, so I could only do it on a part of the image (otherwise during parameter changes I would have to process potentially huge images).

Secondly, the final size of the work is not yet known, and it is on that data that the final rendering actually acts.


> The new UsbSerial2 is much faster than the previous version. But if I create a program in LaserGRBL and then run it in Universal Gcode Sender Classic, it runs twice as fast. [...] The X-axis on my machine moves at a speed of 1000mm / s, but during the execution of the program the speed is much slower.

Speaking of pure stream performance LaserGRBL is able to stream up to 1800 gcode lines/sec to a standard Atmega328, and an even higher value on a ESP32. This is an incredibly fast streaming speed and enough for all classic uses. Keep in mind that the controller doesn't normally execute all of these codes in such a short time, because each command has a time to execute.

Once sent to the controller and guaranteed that the reception buffer is always "well loaded" (the buffer must always be full, otherwise the machine must slow down not knowing what next commands will arrive)  it is only the machine's task to perform the movements at the required speeds. Once the command has been sent, lasergrbl no longer has any influence on the speed with which the machine executes it.

Make sure you have "buffered" and "ultra fast" as settings.

![image](https://user-images.githubusercontent.com/8782035/107935298-805d6200-6f81-11eb-8268-6aadcd9a14ce.png)

**Reply by DevEvV** [reply]
Yes, it's a mix between dithering and grayscale. We do not touch the points where the laser does not work and only change the brightness of the dithering points depending on how bright they are in the original image.
Here is an example of two portraits engraved on the Korelian gabro. This stone is gray and has many stains on it. On basalt or black granite, the image is much cleaner.
I do not have a laser tube at the moment, I ordered it and will be able to send examples in a few weeks. But you can try this engraving method yourself.
I attach the original image, dithering and fake dithering (open at 1: 1 scale). Stone size 120x60 cm, 40 pixels / centimeter.



![IMG_20210215_154831](https://user-images.githubusercontent.com/78816441/108419559-b893c780-7243-11eb-89f3-e1e9b1b1a06a.jpg)
![IMG_20210215_160609](https://user-images.githubusercontent.com/78816441/108419541-b5004080-7243-11eb-9da2-770072c7b1a5.jpg)
![3](https://user-images.githubusercontent.com/78816441/108419578-bf223f00-7243-11eb-9e01-edc32dc704f0.jpg)
![1](https://user-images.githubusercontent.com/78816441/108419591-c21d2f80-7243-11eb-92ff-ffda7512bf82.jpg)
![2](https://user-images.githubusercontent.com/78816441/108419595-c34e5c80-7243-11eb-9a9e-64ce66e14469.jpg)

**Reply by Stevyjb** [reply]
Interesting post, What Dithering mode do you use?

**Reply by DevEvV** [reply]
> Interesting post, What Dithering mode do you use?

I use Floyd-Steinberg dithering and then overlay the original image in multiply mode with 80% opacity.
